IOPool/Input/src/RootPrimaryFileSequence.cc

```diff
@@ -71,24 +71,47 @@ namespace edm {
 
   void RootPrimaryFileSequence::endJob() { closeFile_(); }
 
-  std::unique_ptr<FileBlock> RootPrimaryFileSequence::readFile_() {
+  std::shared_ptr<FileBlock> RootPrimaryFileSequence::readFile_() {
+    std::shared_ptr<FileBlock> fileBlock;
     if (firstFile_) {
-      // The first input file has already been opened.
       firstFile_ = false;
+      // Usually the first input file will already be open
       if (!rootFile()) {
         initFile(input_.skipBadFiles());
       }
+    } else if (goToEventInNewFile_) {
+      goToEventInNewFile_ = false;
+      setAtFileSequenceNumber(goToFileSequenceOffset_);
+      initFile(false);
+      assert(rootFile());
+      bool found = rootFile()->goToEvent(goToEventID_);
+      assert(found);
+    } else if (skipIntoNewFile_) {
+      skipIntoNewFile_ = false;
+      setAtFileSequenceNumber(skipToFileSequenceNumber_);
+      initFile(false);
+      assert(rootFile());
+      if (skipToOffsetInFinalFile_ < 0) {
+        rootFile()->setToLastEntry();
+      }
+      bool atEnd = rootFile()->skipEvents(skipToOffsetInFinalFile_);
+      assert(!atEnd && skipToOffsetInFinalFile_ == 0);
```

Comment:
I am trying to decide if I need to add more comments in the code to explain what is happening here. What do you think? This is complicated.

The first thing is that the existing code is in my opinion incorrect. Probably in the context in which has been used it does not matter and does not cause a problem, but in principle it is wrong.

There are two cases where the existing skipEvents function is used.

Case 1: At the beginning. When we start processing (or do a rewind), we want to skip to the desired event and start processing there. We only want the begin file transitions for the file that contains the current event after skipping. We do not want any end file transitions. This was all already working properly. Plus, it is simple because we are always skipping forward in this case.

Case 2: We are in the middle of processing. Here we want the end file transition for the file we are in before the skip and the begin file transition for the file we are in after the skip. This is a general function available for use in a lot of contexts and I maintained that generality (although in practice, the only actual use case is skipEvents(-2) from EventProcessor.cc when the event display skips back one event).

The first thing I did is to split the single function into two functions, one for each of the above cases. The first function is very simple and does exactly what it was doing before other than the fact it can only skip forward.
I called the new function for case 1 "skipEventsAtBeginning". For case 2, I used the existing name "skipEvents".

Case 1 is doing exactly what it was doing before and does not need further discussion.

There are problems in case 2 in the existing code. First of all, the begin and end file transitions are not being executed. The code is closing and opening files far down inside RootInputFileSequence. To work properly, these transitions need to be run from a very high level in EventProcessor.cc and TransitionProcessors.icc. In the existing code, EventProcessor is not aware that files are being opened and closed and that we end up in a different file.

ProcessBlock depends on these high level transitions. If we want the ProcessBlock transitions to run, then the transitions are needed. We need ProcessBlock transitions if we want access to ProcessBlock data. This is what dragged this issue into this PR.

There is another issue. The FileBlock contains raw pointers to the TTrees in the input file. The EventProcessor contains a smart pointer to the FileBlock. Before this was a unique_ptr and I changed it to a shared_ptr. If RootInputFileSequence opens and closes input  files, then the TTree pointers in the FileBlock become stale. And the FileBlock gets passed around in the transition doRespondToCloseInputFile(FileBlock const& fb). So it is not good for them to be stale (although I couldn't find anything currently using them when I looked).

In the new version of skipEvents, the top part of the function handles skipping in the same file and also the case when there are not any files to go to beyond the original file. This part behaves identically to what existed before.

Then the new skipEvent handles the case when we need to close the current file and open 1 or more other files to find the event we want to skip to. In the new code in this PR, the original state (file and location in the file) is saved and at the end the state of the RootPrimaryFileSequence is returned the original state, except new member variables are set that record the fact that a skip is needed and the file and location in the file we need to skip to. The FileBlock is updated via a new shared_ptr. Then, the PoolSource getNextItemType function drives the EventProcessor through the correct end file and begin file transitions to get to the target event.

Finally, I will note that there is some inefficiency in all this. When a file transition is necessary in the new skipEvent function, the first file close and last file open will occur twice. I didn't like that, but given this is only happening in cmsShow when navigating to a previous event in a different file, it seemed tolerable. It is probably significantly more work to remove that inefficiency.

There are similar issues in the goToEvent which follows skipEvent. It is also only used by cmsShow.